## Условие 1: 1 задача

Скачать файл в уроке\
Считать данные с помощью pandas\
Вывести на экран первые 5 строк\
Посмотреть на описание признаков и на их содержание

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('kc_house_data.csv')

In [3]:
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7      1180.0              0   
1      7242     2.0           0     0  ...      7      2170.0            400   
2     10000     1.0           0     0  ...      6       770.0              0   
3      5000     1.0           0     0  ...      7      1050.0            910   
4      8080     1.0           0     0  ...      8      1680.0              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

В данной таблице представлена информация о продающихся домах, с указанием идентификационного номера, даты, цены, количества спальней, количества ванных комнат (почему количество ванн с плавающей точкой не совсем ясно), жилой площади, площади общей, этажей, с видом ли на воду, разных видов из окна по рейтингу, состояния, оценки, площади над поверхностью, площади подвала, года постройки, года ремонта, почтового индекса, широты, долготы, жилая площадь дома 15ти ближайщих соседей в среднем, общая площадь дома 15ти ближайщих соседей в среднем,

## Условие 2: 2 задача


Проведите первичный анализ данных\
Изучите типы данных\
Найдите количество пропущенных ячеек в данных\
Посчитайте основные статистики по всем признакам и поизучайте их\
Пишите выводы

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21611 non-null  float64
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

In [5]:
df.describe().applymap(lambda x: f'{x:0.2f}')

id       price  bedrooms bathrooms sqft_living    sqft_lot  \
count       21613.00    21613.00  21613.00  21613.00    21613.00    21613.00   
mean   4580301520.86   540088.14      3.37      2.11     2079.90    15106.97   
std    2876565571.31   367127.20      0.93      0.77      918.44    41420.51   
min       1000102.00    75000.00      0.00      0.00      290.00      520.00   
25%    2123049194.00   321950.00      3.00      1.75     1427.00     5040.00   
50%    3904930410.00   450000.00      3.00      2.25     1910.00     7618.00   
75%    7308900445.00   645000.00      4.00      2.50     2550.00    10688.00   
max    9900000190.00  7700000.00     33.00      8.00    13540.00  1651359.00   

         floors waterfront      view condition     grade sqft_above  \
count  21613.00   21613.00  21613.00  21613.00  21613.00   21611.00   
mean       1.49       0.01      0.23      3.41      7.66    1788.40   
std        0.54       0.09      0.77      0.65      1.18     828.13   
min        1.00       0.00      0.00      1.00      1.00     290.00   
25%        1.00       0.00      0.00      3.00      7.00    1190.00   
50%        1.50       0.00      0.00      3.00      7.00    1560.00   
75%        2.00       0.00      0.00      4.00      8.00    2210.00   
max        3.50       1.00      4.00      5.00     13.00    9410.00   

      sqft_basement  yr_built yr_renovated   zipcode       lat      long  \
count      21613.00  21613.00     21613.00  21613.00  21613.00  21613.00   
mean         291.51   1971.01        84.40  98077.94     47.56   -122.21   
std          442.58     29.37       401.68     53.51      0.14      0.14   
min            0.00   1900.00         0.00  98001.00     47.16   -122.52   
25%            0.00   1951.00         0.00  98033.00     47.47   -122.33   
50%            0.00   1975.00         0.00  98065.00     47.57   -122.23   
75%          560.00   1997.00         0.00  98118.00     47.68   -122.12   
max         4820.00   2015.00      2015.00  98199.00     47.78   -121.31   

      sqft_living15 sqft_lot15  
count      21613.00   21613.00  
mean        1986.55   12768.46  
std          685.39   27304.18  
min          399.00     651.00  
25%         1490.00    5100.00  
50%         1840.00    7620.00  
75%         2360.00   10083.00  
max         6210.00  871200.00

На основании первичного анализа данных можно сделать вывод, что таблица состоит из 21613 строк и 21 столбцов, в показателе sqft_above присутсвуют две пустые ячейки. Кроме того практически все данные, кроме даты, являтся числовыми.\
Средняя цена дома составляет 540 088,14$, минимальная 75 000$, максимальная 7 700 тыс.$. Среднее количество спален 3,37, ванных 2,11.\
Минимальня общая площадь 250кв.футов, максимальная 1 651 тыс.кв.футов. Средняя общая площадь 15 106,97 кв.футов, жилая 2 079,9 кв. футов.\
Среднее количество этажей 1,49, максимум 3,5 этажа.\
В среднем дома 1971 годов постройки, самый старый 1900 года, самый новый 2015 года.

## Условие 3: 3 задача

### Ответьте на несколько вопросов:
### 3.1 В каком диапазоне изменяются стоимости недвижимости?

Исходя из основной сводной статистики, цена недвижимости находится в диапазоне от 75 000$ до  7 700 000$.

### 3.2 Какую долю в среднем занимают жилая площадь от всей площади по всем домам?

In [15]:
round(df['sqft_living'].mean()/df['sqft_lot'].mean(),3)

0.138

Доля жилой площади от общей в среднем 13,8% по всем домам на продаже

### 3.3 Как много домов с разными этажами в данных?

In [8]:
df['floors'].value_counts(normalize=True)*100

floors
1.0    49.414704
2.0    38.129829
1.5     8.837274
3.0     2.836256
2.5     0.744922
3.5     0.037015
Name: proportion, dtype: float64

Всего присутствует 6 вариантов домов по этажности. Преобладают одноэтажные дома (49,41%) и двухэтажные (38,12%)

### 3.4 Насколько хорошие состояния у домов в данных?

In [10]:
df['condition'].value_counts(normalize=True)*100

condition
3    64.919262
4    26.275852
5     7.870263
2     0.795817
1     0.138805
Name: proportion, dtype: float64

### 3.5 Найдите года, когда построили первый дом, когда построили последний дом в данных?

In [12]:
df.sort_values('yr_built', ascending=False)

id             date      price  bedrooms  bathrooms  \
21369   923059259  20150401T000000   455950.0         4       2.50   
1763   1832100030  20140625T000000   597326.0         4       4.00   
21262  5694500840  20141125T000000   559000.0         2       3.00   
14215  8156600210  20150326T000000  1285000.0         5       3.50   
20852  1257201420  20140709T000000   595000.0         4       3.25   
...           ...              ...        ...       ...        ...   
8900    917000300  20140520T000000   452000.0         4       1.00   
4438   4232902615  20150428T000000   819000.0         3       1.00   
4397   2767602490  20140724T000000   551000.0         3       1.00   
2566   3037200060  20140926T000000   499000.0         3       2.50   
11266  9827700105  20140917T000000   549000.0         3       2.00   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
21369         2720      5771     2.0           0     0  ...      8   
1763          3570      8250     2.0           0     0  ...     10   
21262         1650       960     3.0           0     0  ...      8   
14215         2980      5100     2.0           0     0  ...     10   
20852         3730      4560     2.0           0     0  ...      9   
...            ...       ...     ...         ...   ...  ...    ...   
8900          1210      3760     1.5           0     0  ...      6   
4438          1300      3600     2.0           0     0  ...      7   
4397           940      1948     1.0           0     0  ...      6   
2566          1750      2150     2.5           0     0  ...      7   
11266         2330      3600     1.5           0     0  ...      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
21369      2720.0              0      2015             0    98056  47.4917   
1763       2860.0            710      2015             0    98040  47.5784   
21262      1350.0            300      2015             0    98103  47.6611   
14215      2370.0            610      2015             0    98115  47.6782   
20852      2760.0            970      2015             0    98103  47.6725   
...           ...            ...       ...           ...      ...      ...   
8900       1210.0              0      1900             0    98103  47.6872   
4438       1300.0              0      1900             0    98119  47.6345   
4397        940.0              0      1900             0    98107  47.6733   
2566       1230.0            520      1900          2014    98122  47.6037   
11266      1580.0            750      1900             0    98122  47.6025   

          long  sqft_living15  sqft_lot15  
21369 -122.170           1940        4184  
1763  -122.226           2230       10000  
21262 -122.346           1650        3000  
14215 -122.299           1780        5100  
20852 -122.330           1800        4560  
...        ...            ...         ...  
8900  -122.344           1540        3800  
4438  -122.366           2510        4800  
4397  -122.383           1700        5000  
2566  -122.311           1410        3300  
11266 -122.303           1750        3600  

[21613 rows x 21 columns]

Самый первый дом построен в 1900 году, самый последний в 2015 году

## Условие 4: 4 задача
## Ответьте на несколько вопросов

### 4.1 Сколько в среднем стоят дома, у которых 2 спальни?

In [9]:
df[df['bedrooms']==2]['price'].mean()

401372.681884058

### 4.2 Какая в среднем общая площадь домов, у которых стоимость больше 600 000?

In [6]:
df[df['price']>=600000]['sqft_lot'].mean()

20454.783537543197

### 4.3 Как много домов коснулся ремонт?

In [9]:
df[~(df['yr_renovated']==0)]

id             date      price  bedrooms  bathrooms  \
1      6414100192  20141209T000000   538000.0         3       2.25   
35     9547205180  20140613T000000   696000.0         3       2.50   
95     1483300570  20140908T000000   905000.0         4       2.50   
103    2450000295  20141007T000000  1088000.0         3       2.50   
115    3626039325  20141121T000000   740500.0         3       3.50   
...           ...              ...        ...       ...        ...   
19622  7351200295  20150114T000000  1150000.0         3       1.75   
20057   126039256  20140904T000000   434900.0         3       2.00   
20444  4305600360  20150225T000000   500012.0         4       2.50   
20447  3319500628  20150212T000000   356999.0         3       1.50   
20962  1278000210  20150311T000000   110000.0         2       1.00   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
1             2570      7242     2.0           0     0  ...      7   
35            2300      3060     1.5           0     0  ...      8   
95            3300     10250     1.0           0     0  ...      7   
103           2920      8113     2.0           0     0  ...      8   
115           4380      6350     2.0           0     0  ...      8   
...            ...       ...     ...         ...   ...  ...    ...   
19622         1760      6788     2.0           1     4  ...      7   
20057         1520      5040     2.0           0     0  ...      7   
20444         2400      9612     1.0           0     0  ...      8   
20447         1010      1546     2.0           0     0  ...      8   
20962          828      4524     1.0           0     0  ...      6   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
1          2170.0            400      1951          1991    98125  47.7210   
35         1510.0            790      1930          2002    98115  47.6827   
95         2390.0            910      1946          1991    98040  47.5873   
103        2920.0              0      1950          2010    98004  47.5814   
115        2780.0           1600      1900          1999    98117  47.6981   
...           ...            ...       ...           ...      ...      ...   
19622      1760.0              0      1940          1960    98125  47.7336   
20057      1520.0              0      1977          2006    98177  47.7770   
20444      1230.0           1170      1962          2009    98059  47.4799   
20447      1010.0              0      1971          2014    98144  47.5998   
20962       828.0              0      1968          2007    98001  47.2655   

          long  sqft_living15  sqft_lot15  
1     -122.319           1690        7639  
35    -122.310           1590        3264  
95    -122.249           1950        6045  
103   -122.196           2370        8113  
115   -122.368           1830        6350  
...        ...            ...         ...  
19622 -122.284           1630        7588  
20057 -122.362           1860        8710  
20444 -122.127           2430        5539  
20447 -122.311           1010        1517  
20962 -122.244            828        5402  

[914 rows x 21 columns]

914 домов было отремонтировано

### 4.4 Насколько в среднем стоимость домов с оценкой grade домов выше 10 отличается от стоимости домов с оценкой grade меньше 4?

In [12]:
df_grade10 = df[df['grade']>10]['price'].mean()
df_grade10

1257984.857579462

In [13]:
df_grade4 = df[df['grade']<4]['price'].mean()
df_grade4

189750.0

In [14]:
df_grade10-df_grade4

1068234.857579462

Средняя стоимость домов с оценкой grade выше 10 отличается от стоимости домов с оценкой grade меньше 4 на 1 068 234,85$

## Условие 5: 5 задача
## Ответьте на несколько вопросов

### 5.1 Выберите дом клиенту
### Клиент хочет дом с видом на набережную, как минимум с тремя ванными и с подвалом. Сколько вариантов есть у клиента?

In [15]:
df[~(df['waterfront']==0)&(df['bedrooms']>=3)&(df['sqft_basement']>0)]

id             date      price  bedrooms  bathrooms  \
49      822039084  20150311T000000  1350000.0         3       2.50   
246    2025069065  20140929T000000  2400000.0         4       2.50   
300    3225069065  20140624T000000  3075000.0         4       5.00   
457    2122039094  20141126T000000   705000.0         3       3.00   
656    3760500116  20141120T000000  3070000.0         3       2.50   
...           ...              ...        ...       ...        ...   
18914  9253900408  20150408T000000  1400000.0         3       2.75   
20325   518500480  20140811T000000  3000000.0         3       3.50   
20767  8043700105  20150417T000000  2300000.0         4       4.00   
21201   518500460  20141008T000000  2230000.0         3       3.50   
21576  9253900271  20150107T000000  3567000.0         5       4.50   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
49            2753     65005     1.0           1     2  ...      9   
246           3650      8354     1.0           1     4  ...      9   
300           4550     18641     1.0           1     4  ...     10   
457           1970     20978     2.0           1     3  ...      9   
656           3930     55867     1.0           1     4  ...      8   
...            ...       ...     ...         ...   ...  ...    ...   
18914         3130     19530     1.0           1     4  ...      8   
20325         4410     10756     2.0           1     4  ...     11   
20767         4360      8175     2.5           1     4  ...     10   
21201         3760      5634     2.0           1     4  ...     11   
21576         4850     10584     2.0           1     4  ...     10   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
49         2165.0            588      1953             0    98070  47.4041   
246        1830.0           1820      2000             0    98074  47.6338   
300        2600.0           1950      2002             0    98074  47.6053   
457        1770.0            200      1980             0    98070  47.3844   
656        2330.0           1600      1957             0    98034  47.7022   
...           ...            ...       ...           ...      ...      ...   
18914      1690.0           1440      1947          1984    98008  47.5895   
20325      3430.0            980      2014             0    98056  47.5283   
20767      3940.0            420      2007             0    98008  47.5724   
21201      2830.0            930      2014             0    98056  47.5285   
21576      3540.0           1310      2007             0    98008  47.5943   

          long  sqft_living15  sqft_lot15  
49    -122.451           2680       72513  
246   -122.072           3120       18841  
300   -122.077           4550       19508  
457   -122.438           2280       75396  
656   -122.224           2730       26324  
...        ...            ...         ...  
18914 -122.111           2980       18782  
20325 -122.205           3550        5634  
20767 -122.104           2670        8525  
21201 -122.205           3560        5762  
21576 -122.110           3470       18270  

[84 rows x 21 columns]

По данному запросу присутсвует 84 варианта

### 5.2 Выберите дом клиенту
### Клиент хочет дом либо с очень красивым видом из окна, либо с видом на набережную, в очень хорошем состоянии и год постройки не меньше 1980 года. В какой ценовом диапазоне будут дома?

In [22]:
sorted(df['view'].unique())

[0, 1, 2, 3, 4]

In [23]:
sorted(df['waterfront'].unique())

[0, 1]

In [24]:
sorted(df['grade'].unique())

[1, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13]

In [27]:
df[((df['view']==4)|(df['waterfront']==1))&(df['grade']>10)&(df['yr_built']>=1980)].sort_values('price')

id             date      price  bedrooms  bathrooms  \
8568   8653600100  20150330T000000   750000.0         5       2.50   
21050  2424059170  20150219T000000   900000.0         5       6.00   
10665  7852160310  20140814T000000  1010000.0         4       2.75   
2713   7851980260  20140730T000000  1110000.0         5       3.50   
20923  7852070210  20140527T000000  1149000.0         4       3.00   
12429    98001070  20140818T000000  1169000.0         5       4.25   
10385  3225069301  20140612T000000  1228000.0         4       2.50   
16260  8965510190  20140610T000000  1250000.0         4       2.50   
16837  4441300170  20150112T000000  1300000.0         4       2.50   
13770  1795800040  20140903T000000  1350000.0         4       3.25   
20005   148000475  20140528T000000  1400000.0         4       3.25   
14774  9412400220  20140710T000000  1612500.0         4       2.75   
7269   4139470010  20141006T000000  1615000.0         4       3.25   
8174   9279700150  20150212T000000  1625000.0         4       3.75   
15668  4030100290  20141001T000000  1680000.0         5       3.50   
13068  7856410411  20140922T000000  1698890.0         4       4.50   
11829  4139420640  20141030T000000  1785000.0         4       3.50   
5749   6744701310  20150415T000000  1850000.0         4       2.50   
3157   1827200265  20140911T000000  1899000.0         2       2.75   
6501   2626069030  20150209T000000  1940000.0         4       5.75   
7700   3761700053  20150105T000000  2150000.0         3       2.75   
6233   2024069008  20140619T000000  2200000.0         5       4.75   
21201   518500460  20141008T000000  2230000.0         3       3.50   
12777  1225069038  20140505T000000  2280000.0         7       8.00   
11402  3426049284  20140819T000000  2300000.0         4       3.25   
4013    724069059  20140509T000000  2400000.0         3       2.25   
15828  4139420190  20150512T000000  2480000.0         4       5.00   
17558  7631800015  20150407T000000  2510000.0         3       3.25   
16817   825059178  20140923T000000  2574000.0         4       3.75   
18455  8043700300  20140608T000000  2700000.0         4       3.25   
4190   2303900090  20140729T000000  2880500.0         4       2.50   
20325   518500480  20140811T000000  3000000.0         3       3.50   
7313   4131900066  20140825T000000  3100000.0         3       3.00   
12872  2424049029  20140529T000000  3100000.0         6       4.25   
10465   333100295  20141124T000000  3120000.0         3       3.50   
13528  3025059124  20140828T000000  3168750.0         5       3.50   
11535  8964800890  20150109T000000  3200000.0         3       3.25   
15482   624069108  20140812T000000  3200000.0         4       3.25   
2444   7524900003  20141210T000000  3278000.0         2       1.75   
14083  3625059043  20140904T000000  3300000.0         5       4.75   
15255  2425049063  20140911T000000  3640900.0         4       3.25   
19017  2303900100  20140911T000000  3800000.0         3       4.25   
2085   8106100105  20141114T000000  3850000.0         4       4.25   
2626   7738500731  20140815T000000  4500000.0         5       5.50   
8092   1924059029  20140617T000000  4668000.0         5       6.75   
1164   1247600105  20141020T000000  5110800.0         5       5.25   
1315   7558700030  20150413T000000  5300000.0         6       6.00   
1448   8907500070  20150413T000000  5350000.0         5       5.00   
9254   9208900037  20140919T000000  6885000.0         6       7.75   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
8568          3120     15593     2.0           0     4  ...     11   
21050         7120     40806     2.0           0     4  ...     12   
10665         3430     15877     1.0           0     4  ...     11   
2713          7350     12231     2.0           0     4  ...     11   
20923         5940     11533     2.0           0     4  ...     11   
12429         4610     13252     2.0           0     4  ...     11   
10385         5730     4

Диапазон цен составит: от 750 000$ до 6 885 000$

### 5.3 Выберите дом клиенту
### Клиент хочет дом без подвала, с двумя этажами, стоимостью до 150000. Какая оценка по состоянию у таких домов в среднем?

In [28]:
df[(df['sqft_basement']==0)&(df['floors']==2)&(df['price']<150000)]['grade'].mean()

7.0

В среднем 7 баллов из 13